apps/managedidentity/managedidentity.go

```diff
@@ -129,60 +168,108 @@ func New(id ID, options ...ClientOption) (Client, error) {
 	default:
 		return Client{}, fmt.Errorf("unsupported type %T", id)
 	}
+
 	client := Client{
 		miType:     id,
 		httpClient: opts.httpClient,
+		source:     source,
 	}
 
 	return client, nil
 }
 
-func createIMDSAuthRequest(ctx context.Context, id ID, resource string, claims string) (*http.Request, error) {
-	var msiEndpoint *url.URL
-	msiEndpoint, err := url.Parse(imdsEndpoint)
-	if err != nil {
-		return nil, fmt.Errorf("couldn't parse %q: %s", imdsEndpoint, err)
+// Detects and returns the managed identity source available on the environment.
+func GetSource(id ID) (Source, error) {
+	identityEndpoint := os.Getenv(IdentityEndpointEnvVar)
+	identityHeader := os.Getenv(IdentityHeaderEnvVar)
+	identityServerThumbprint := os.Getenv(IdentityServerThumbprintEnvVar)
+	msiEndpoint := os.Getenv(MsiEndpointEnvVar)
+
+	if identityEndpoint != "" && identityHeader != "" {
+		if identityServerThumbprint != "" {
+			return ServiceFabric, nil
+		}
+		return AppService, nil
+	} else if msiEndpoint != "" {
+		return CloudShell, nil
+	} else if _, isValid := getAndValidateAzureArcEnvVars(); isValid {
+		return AzureArc, nil
 	}
-	msiParameters := msiEndpoint.Query()
-	msiParameters.Set(apiVersionQuerryParameterName, imdsAPIVersion)
-	resource = strings.TrimSuffix(resource, "/.default")
-	msiParameters.Set(resourceQuerryParameterName, resource)
 
-	if len(claims) > 0 {
-		msiParameters.Set("claims", claims)
+	return DefaultToIMDS, nil
+}
+
+// Acquires tokens from the configured managed identity on an azure resource.
+//
+// Resource: scopes application is requesting access to
+// Options: [WithClaims]
+func (client Client) AcquireToken(ctx context.Context, resource string, options ...AcquireTokenOption) (base.AuthResult, error) {
+	o := AcquireTokenOptions{}
+	var req *http.Request
+	var err error
+
+	for _, option := range options {
+		option(&o)
 	}
+	var tokenResponse accesstokens.TokenResponse
+	switch client.source {
+	case AzureArc:
+		req, err = createAzureArcAuthRequest(ctx, client.miType, resource)
+		if err != nil {
+			return base.AuthResult{}, err
+		}
 
-	switch t := id.(type) {
-	case UserAssignedClientID:
-		msiParameters.Set(miQueryParameterClientId, string(t))
-	case UserAssignedResourceID:
-		msiParameters.Set(miQueryParameterResourceId, string(t))
-	case UserAssignedObjectID:
-		msiParameters.Set(miQueryParameterObjectId, string(t))
-	case systemAssignedValue: // not adding anything
+		// need to perform preliminary request to retrieve the secret key challenge provided by the HIMDS service
+		// this is done when we get a 401 response, which will be handled by the response handler
+		tokenResponse, err = client.getTokenForRequest(ctx, req)
+		if err != nil {
+			switch callErr := err.(type) {
+			case errors.CallErr:
```

Comment:
It's strange to implement the expected success case as error handling. I believe a more straightforward approach [like this](https://github.com/Azure/azure-sdk-for-go/blob/7a626d2e848f739e4f6f94b275f1f1b2eeb56300/sdk/azidentity/managed_identity_client.go#L425) would be easier to understand and maintain.

If you want to keep this approach, please at least use [errors.As](https://pkg.go.dev/errors#As) to make it more robust

Reply:
Opted to use errors.As